src/utils/createReadStreamInternal.ts

```diff
@@ -0,0 +1,431 @@
+import {GetRowsOptions} from '../table';
+import {Row} from '../row';
+import * as is from 'is';
+import {Filter, BoundData} from '../filter';
+import {Mutation} from '../mutation';
+import {AbortableDuplex} from '../index';
+import {
+  ChunkPushData,
+  ChunkPushLastScannedRowData,
+  ChunkTransformer,
+  DataEvent,
+} from '../chunktransformer';
+import {TableUtils} from './table';
+import {Duplex, PassThrough, Transform} from 'stream';
+import {
+  MethodName,
+  StreamingState,
+} from '../client-side-metrics/client-side-metrics-attributes';
+import {google} from '../../protos/protos';
+const pumpify = require('pumpify');
+import {grpc, ServiceError} from 'google-gax';
+import {
+  DEFAULT_BACKOFF_SETTINGS,
+  getNextDelay,
+  IGNORED_STATUS_CODES,
+  populateAttemptHeader,
+  RETRYABLE_STATUS_CODES,
+  TabularApiSurface,
+} from '../tabular-api-surface';
+
+/**
+ * Creates a readable stream of rows from a Bigtable table or authorized view.
+ *
+ * This internal method handles the core logic for streaming rows from a Bigtable
+ * table. It supports various filtering, limiting, and retry mechanisms. It can
+ * be used to create a stream for either a whole table or an authorized view.
+ *
+ * @param {Table} table The Table instance to read rows from.
+ * @param {boolean} singleRow boolean to check if the request is for a single row.
+ * @param {GetRowsOptions} [opts] Optional configuration for the read operation.
+ * @param {boolean} [opts.decode=true] If set to `false` it will not decode
+ *     Buffer values returned from Bigtable.
+ * @param {boolean} [opts.encoding] The encoding to use when converting
+ *     Buffer values to a string.
+ * @param {string} [opts.end] End value for key range.
+ * @param {Filter} [opts.filter] Row filters allow you to
+ *     both make advanced queries and format how the data is returned.
+ * @param {object} [opts.gaxOptions] Request configuration options, outlined
+ *     here: https://googleapis.github.io/gax-nodejs/CallSettings.html.
+ * @param {string[]} [opts.keys] A list of row keys.
+ * @param {number} [opts.limit] Maximum number of rows to be returned.
+ * @param {string} [opts.prefix] Prefix that the row key must match.
+ * @param {string[]} [opts.prefixes] List of prefixes that a row key must
+ *     match.
+ * @param {object[]} [opts.ranges] A list of key ranges.
+ * @param {string} [opts.start] Start value for key range.
+ * @param {string} [viewName] The name of the authorized view, if applicable.
+ * @returns {stream} A readable stream of {@link Row} objects.
+ *
+ */
+export function createReadStreamInternal(
```

Comment:
I didn't look at this file in detail except for changes related to `singleRow`. I assume everything is copy pasted from createReadStream. Let me know if there are other lines i should take a look.